src/niri.rs

```diff
@@ -1194,12 +1194,17 @@ impl Niri {
                     .as_ref()
                     .map(|surface| (surface, bbox_from_surface_tree(surface, surface_pos)));
 
+                // FIXME we basically need to pick the largest scale factor across the overlapping
+                // outputs, this is how it's usually done in clients as well.
+                let mut cursor_scale = 1;
+                let mut dnd_scale = 1;
```

Comment:
This doesn't really work, because you run this on every frame meaning that the scale is being send in a busy loop. There're two approaches I can see

1) Cache on per-surface basis the current scale factor and don't resent it.
2) Send scale factor for the pointer once it `enters` surface, given that surface
can't span multiple outputs.

Which approach should we pick?

Though, I wonder whether we should send this in reply to `surface::commit` when the surface is a cursor surface? And then we need to resend scale when output actually got changed for the cursor, but I'm not sure how to identify that with the smithay's API.

Reply:
I think caching the current scale factor per-surface and not re-sending is what I'd prefer here.

> Send scale factor for the pointer once it enters surface, given that surface
can't span multiple outputs.

This doesn't quite work in spirit, because the cursor (or, better example, the DnD surface) can be large and significantly overlap another output with a higher scale.

> Though, I wonder whether we should send this in reply to surface::commit when the surface is a cursor surface? And then we need to resend scale when output actually got changed for the cursor, but I'm not sure how to identify that with the smithay's API.

Yeah, I'm not sure Smithay has API for this atm. I think if we go with the caching approach it should work fine as is.

Reply:
The new smithay API should at least cache events we're sending. So caching is now done globally.